data_preparation/AgML_predictor_data_prepare.py

```diff
@@ -0,0 +1,1077 @@
+           touched by geom).
+        2. we get the geometry window bounds of the lowest resolution dataset and align them to the highest
+           resolution dataset.
+            - lowest resolution dataset window should be the maximum extent, extents covering all other dataset
+              windows.
+            - We align the bounds to the highest resolution dataset by defining the bounds window on the hres
+              dataset and rounding them to the closest whole pixel
+        3. we define unified read window and out_shape
+            - We use the aligned bounds to define a read window for all datasets
+            - We use the resolution of the hres rounded bounds window shape as the read out_shape
+    :param geom: GeoJSON-like feature (implements __geo_interface__) – feature collection, or geometry.
+    :param ds_list: List of raster file paths or rasterio datasets
+    :return: Pair of tuples, shape and bounds - ((x_min, y_min, x_max, y_max), (rows, columns))
+    """
+    max_res_ds = None
+    max_res_window_size = None
+    min_res_window_size = None
+    max_extent_window_bounds = None
+
+    for ds in ds_list:
+        dataset = ds if isinstance(ds, rasterio.DatasetReader) else rasterio.open(ds)
+        # dataset = ds.to_rio()
+        ds_window = rasterio.features.geometry_window(dataset, geom)
+
+        if not max_res_ds:  # just assign values to all vars if first file in the loop
+            max_res_ds = dataset
+            max_res_window_size = rasterio.windows.shape(ds_window)
+            min_res_window_size = rasterio.windows.shape(ds_window)
+            max_extent_window_bounds = dataset.window_bounds(ds_window)
+        elif rasterio.windows.shape(ds_window) > max_res_window_size:
+            max_res_ds = dataset
+            max_res_window_size = rasterio.windows.shape(ds_window)
+        elif rasterio.windows.shape(ds_window) < min_res_window_size:
+            max_extent_window_bounds = dataset.window_bounds(ds_window)
+            min_res_window_size = rasterio.windows.shape(ds_window)
+
+    # new bounds fitted to the highest resolution file
+    out_bounds = max_res_ds.window_bounds(round_window(max_res_ds.window(*max_extent_window_bounds)))
+    out_shape = rasterio.windows.shape(round_window(max_res_ds.window(*out_bounds)))
+
+    return out_bounds, out_shape
+
+"""
+@author: Joint Research Center
+"""
+def arr_stats(arr, weights=None,
+              output=('min', 'max', 'sum', 'mean', 'count', 'mode')):
+    """
+    Extracts statistics from input array (arr).
+    It uses weights array as weights if provided.
+    List of statistics to extract is defined in the output parameter as a list or a comma separated string.
+    :param arr: Input array
+    :param weights: Array providing weights for each pixel value. Used to calculate stats.
+    :param output: List of values to extract, can be a list or comma separated string.
+        Possible values are:
```

Comment:
Missing description for output option "mode"

Reply:
@umdsgy, @krsnapaudel after another colleague asked for something similar I realised that this implementation doesn't take into account the AFI. The output would be correct only if it's done with no AFI provided.